Improve performance avoiding unnecessary vos dumps

marciobarbosa opened the pull request:
Currently, every single job is completely independent of each other. As
a result, when a volume is being synchronized between several destination
cells, the same volume is dumped multiple times. In order to improve the
performance, this patch introduces a set of changes aiming to avoid
unnecessary time consuming dumps. To make it possible, the jobs
requesting the same volume with the same timestamp are grouped into a
hash table and only one dump file is generated per entry. The dump file
for a specific entry is assigned to the first job of the list and the
remaining jobs receive a copy of the file in question.

adeason commented on lib/AFS/CellCC/Dump.pm:
```diff
@@ -166,80 +173,83 @@ _calc_incremental($$) {
+    my @dump_fhs;
+    for my $job (@jobs) {
+        my $dump_fh = File::Temp->new(DIR => config_get('dump/scratch-dir'),
+                                      TEMPLATE => "cccdump_job$job->{jobid}_XXXXXX",
```
It might be nice to have the "parent" jobid in here, too. Like, "cccdump_job$jobs[0]->{jobid}_job$job->{jobid}_XXXXXX" . It just makes it easier to see what jobid we used when doing the actual dump, if for whatever reason any data between the jobs differ.

adeason commented on lib/AFS/CellCC/Dump.pm:
```diff
@@ -166,80 +173,83 @@ _calc_incremental($$) {
-               description => "Starting to dump volume");
-    $job->{vol_lastupdate} = $lastupdate;
+    my $descr = "Starting to dump volume";
+    for my $i (0 ,, $#jobs) {
```
I think you meant "0 .. $#jobs"? Running "0 ,, $#jobs" yields like (0,5) instead of (0,1,2,3,4,5)

adeason commented on lib/AFS/CellCC/Dump.pm:
```diff
@@ -166,80 +173,83 @@ _calc_incremental($$) {
-    $job->{vol_lastupdate} = $lastupdate;
+    my $descr = "Starting to dump volume";
+    for my $i (0 ,, $#jobs) {
+        if ($i == 1) {
```
IMO this is written more clearly like this is inside the loop:
```
if ($i == 0) {
    $descr = "Starting to dump volume";
} else {
    $descr = "Waiting for dump from job $jobs[0]->{jobid}";
}
```

There's no reason to be clever/fancy here to try to be more efficient; clarity is more important.

adeason commented on lib/AFS/CellCC/Dump.pm:
```diff
@@ -41,35 +41,42 @@ our @EXPORT_OK = qw(process_dumps);
+    seek($dump_fhs[0], 0, 0);
 
+    foreach my $i (1 .. $#dump_fhs) {
+        File::Copy::copy($dump_fhs[0]->filename, $dump_fhs[$i]->filename);
```
I'm not sure I see a reason to do a full file copy? Just using `link($dump_fhs[0]->filename, $dump_fhs[$i]->filename);` will be much faster, and we don't modify the files after this regardless.

adeason commented on lib/AFS/CellCC/Dump.pm:
```diff
@@ -309,32 +351,78 @@ process_dumps($$$@) {
-            next;
+        eval {
+            my $volname = $job->{volname};
+            my $lastupdate = _calc_incremental($job, $start_state);
```
I really really do not want to run `_calc_incremental` directly inside `process_dumps` if we don't have to... that will prevent all other dumps from processing until it finishes running and can let someone else come and "steal" the jobs out from under us; that's why we spawn all of the real work in child procs. But of course we can't just spawn this block in a child proc, because we need to access the global %vol_table. Instead what we can do that I think is still acceptable is create a child proc per volume name, and have that proc run `_calc_incremental` for each job for that volume name. That can just populate the vol_lastupdate field in the db and change the state, and `_process_dumps` can find all of the jobs in that state and populate %vol_table from that data.

Do you think that can work? We'll still serialize `_calc_incremental` calls for different jobs for the same volumes, but jobs for different volumes won't hold each other up.

adeason replied:
>Instead what we can do that I think is still acceptable is create a child proc per volume name

Or no, that doesn't make much sense. If we're going to do this, just create a child proc per job like normal, and have that proc fill in `vol_lastupdate` in the db. Wait until all jobs are done doing that, and just proceed like normal; the parent process can then fill in the `%vol_table` table from data that's in the db.

adeason commented on lib/AFS/CellCC/Dump.pm:
```diff
@@ -41,35 +41,42 @@ our @EXPORT_OK = qw(process_dumps);
+    my $checksum = calc_checksum_jobs($dump_fhs[0], $filesize, $algo, $prev_state, @jobs);
+
+    foreach my $i (0 .. $#jobs) {
+        update_job(jobid => $jobs[$i]->{jobid},
```
Ideally we'd have db primitives that can act on multiple jobs at once, so we don't need to issue separate queries like this. However, even without that this can be made a little more efficient just by wrapping this in a db_rw() call. See `startsync()` in `lib/AFS/CellCC.pm` for an example. That should let us issues these multiple calls in a single db connection and transaction, so it should be impossible to have only some of them fail.

adeason commented on lib/AFS/CellCC/Dump.pm:
```diff
@@ -166,80 +173,83 @@ _calc_incremental($$) {
-        # _calc_incremental said we can skip syncing the volume, so transition
-        # the job straight to the final stage
-        DEBUG "volume $volname appears to not need a sync";
+    for my $job (@jobs) {
```
these `update_job` calls should also be wrapped by `db_rw`

adeason commented on lib/AFS/CellCC/Dump.pm:
```diff
@@ -166,80 +173,83 @@ _calc_incremental($$) {
+        if ($i == 1) {
+            $descr = "Waiting for dump from job $jobs[0]->{jobid}";
+        }
+        update_job(jobid => $jobs[$i]->{jobid},
```
these `update_job` calls should also be wrapped by `db_rw`

adeason commented on lib/AFS/CellCC/Dump.pm:
```diff
@@ -254,40 +264,63 @@ _do_dump($$$) {
-                       timeout => $interval+60,
-                       description => $descr);
+            $descr = "Running vos dump ($pretty_bytes / $pretty_total dumped, $pretty_rate)";
+            for my $i (0 .. $#jobs) {
```
these `update_job` calls should also be wrapped by `db_rw`

updating status for all "child" jobs also seems unnecessary, but I guess we need to make sure they don't timeout... so oh well

adeason commented on lib/AFS/CellCC/Dump.pm:
```diff
@@ -254,40 +264,63 @@ _do_dump($$$) {
-               from_state => $state,
-               timeout => 120,
-               description => "Processing finished dump file");
+    for my $job (@jobs) {
```
db_rw here, too

adeason commented on lib/AFS/CellCC/Dump.pm:
```diff
@@ -254,40 +264,63 @@ _do_dump($$$) {
+    my ($elapsed_time, @jobs) = @_;
 
-    INFO "Finished performing dump for job $job->{jobid} (vol '$job->{volname}', $job->{src_cell} -> $job->{dst_cell})";
+    for my $job (@jobs) {
```
db_rw here too (if we keep this function)

adeason commented on lib/AFS/CellCC/Util.pm:
```diff
@@ -401,4 +463,71 @@ calc_checksum($$$$$$) {
+            $bytes = pretty_bytes($pos);
+            $descr = "Checksumming dump blob ($bytes / $total)";
+
+            for my $job (@jobs) {
```
these `update_job` calls should also be wrapped by `db_rw`

adeason commented on lib/AFS/CellCC/Util.pm:
```diff
@@ -340,6 +340,68 @@ scratch_ok($$$$$) {
+# assuming we should not have that many destination cells, the size of the
+# hardlinks are not taken into consideration.
+sub
+scratch_ok_jobs($$$$@) {
```
having a separate function for this seems really unnecessary. just make scratch_ok take an array of jobs? and if there are still existing callers that only use 1 job, just pass an array only containing that 1 job

adeason reviewed (changes requested):
I think I agree with the general approach here, but we can't be running _calc_incremental without spawning a child proc. Some notes inline on how that can be made to work (along with other minor things)

adeason commented:
Oh also, I forgot to mention: a "nice to have" here would be to split this into two commits: one more mundane commit that converts some existing functions to accept an array of jobs instead of a single job (and updates callers to give an array with just 1 job in it), and a more interesting commit that actually gives those functions multiple jobs at once. Only do that if it doesn't seem difficult; don't worry about it if it creates a lot more work.

marciobarbosa committed:
Refactor a few functions to handle multiple jobs

In preparation for upcoming enhancements, change a few functions to
operate on several jobs at once. This includes _do_dump, scratch_ok,
calc_checksum, and _dump_success.

For _get_size, just remove the $job argument, since it isn't even
currently used.

Currently we only operate on one job at a time, so this commit should
incur no visible differences.

[adeason@sinenomine.net: changed tempfile handling for dump
hardlinks.]

marciobarbosa committed:
Improve performance avoiding unnecessary vos dumps

Currently, every single job is completely independent of each other. As
a result, when a volume is being synchronized between several destination
cells, the same volume is dumped multiple times. In order to improve the
performance, this patch introduces a set of changes aiming to avoid
unnecessary time consuming dumps. To make it possible, the jobs
requesting the same volume with the same timestamp are grouped into a
hash table and only one dump file is generated per entry. The dump file
for a specific entry is assigned to the first job of the list and the
remaining jobs receive a copy of that dump file.

[adeason@sinenomine.net: reworked incremental calculations and
%lu_table population.]

adeason committed:
DB: Allow find_jobs to accept arrayrefs

Alter find_jobs so that it can find jobs for multiple different values
for 'dst_cell' or 'state'. That is, if the caller specifies an
arrayref for 'dst_cell' or 'state', we'll find jobs for any of the
given destination cells and any of the given states.

This also changes find_jobs to return the job state as one of the
returned fields, and lets find_jobs be called without a dbhandle (so
it can be called more easily outside of DB.pm).

Change find_update_jobs to make use of this functionality for
'dst_cell'; future commits will make use of the functionality for
'state'.

adeason commented on lib/AFS/CellCC/Util.pm:
```diff
@@ -401,4 +401,18 @@ calc_checksum($$$$$$) {
+    my ($field, @jobs) = @_;
+    my @info;
+    for my $job (@jobs) {
+        if (exists $job->{$field}) {
```
I don't think you want to silently ignore non-existent fields. Did you see this actually happen? It seems like that would be a serious error, for the places this is used.

adeason commented on lib/AFS/CellCC/Util.pm:
```diff
@@ -401,4 +401,18 @@ calc_checksum($$$$$$) {
+# Get the value of the field provided as an argument for every job and return
+# these values as an array.
+sub
+get_jobs_info($@) {
```
eh... I don't think this is needed. This can be replaced by a pretty simple map call, like: `map { $_->{jobid} } @jobs;` (this is really a comment on the whole commit, but github doesn't seem to let me comment on the commit itself?)

adeason commented on lib/AFS/CellCC/DB.pm:
```diff
@@ -484,6 +485,64 @@ update_job(%) {
 
+# Similar to update_job but updates several jobs at once.
+sub
+update_jobs($%) {
```
I had previously mentioned that we should maybe have a db primitive to update several jobs at once, to avoid those loops in all of the callers. But I meant having another SQL query to do all of the updates at once, not just having a function that calls _job_set in a loop. I don't think this function is worth it; it adds some new code but it's almost the same as just update_job(); I'd rather have the loops in the caller.

Actually having some SQL query do the updates all at once is maybe desirable at some point, but after thinking about it, I don't think it's worth the complexity right now; speed here isn't important right now.

adeason commented on lib/AFS/CellCC/Util.pm:
```diff
@@ -277,23 +277,26 @@ _unpretty_bytes($) {
-               to_state => $state,
-               timeout => undef,
-               description => "Waiting for enough scratch disk space to be free");
+_scratch_rollback($@) {
```
in general, I'd rather these functions keep the order of their arguments the same; just replace `$job` with an arrayref of jobs. I just think it makes it easier to update the arguments in the future (obv we can only have one array arg, but multiple arrayref/hashref args)

adeason commented on lib/AFS/CellCC/Dump.pm:
```diff
@@ -207,9 +207,9 @@ _do_dump($$$) {
+    if (!scratch_ok($prev_state, $dump_size,
                     config_get('dump/scratch-dir'),
-                    config_get('dump/scratch-minfree'))) {
+                    config_get('dump/scratch-minfree'), $job)) {
```
a comment for the whole commit: these refactoring commits I think should be all done as one single commit. They all do the same thing, just for different functions, and keeping them separate leaves some code in this weird "in-between" state where some of these functions take arrays of jobs, and others take single jobs (and some of these are really small/mundane changes). This line, for instance, calls scratch_ok with a single job, but we're never going to actually see that line stay like that, since in the future it's going to change to pass in an array of jobs; the same array that was given to us.

adeason commented on lib/AFS/CellCC/Dump.pm:
```diff
@@ -185,77 +185,82 @@ _calc_incremental($$) {
 sub
-_do_dump($$$) {
-    my ($server, $job, $prev_state) = @_;
+_do_dump($$@) {
```
the changes to `_do_dump` I think should be moved to the "multiple jobs refactoring" changes, not in here; many of the changes here are just to make `_do_dump` handle multiple jobs at once.

adeason commented on lib/AFS/CellCC/Dump.pm:
```diff
@@ -185,77 +185,82 @@ _calc_incremental($$) {
+    my @dump_fhs;
+    my $pjob = $jobs[0]->{jobid};
+    for my $job (@jobs) {
+        my $dump_fh = File::Temp->new(DIR => config_get('dump/scratch-dir'),
```
I don't think there's a reason to create these tempfiles so early. it's not a big deal, but it means these temp files stay around for longer while they're not needed (which means a higher chance of random crud being left around, since these files aren't recorded in the database yet). we could leave this as normal, and create the temp files for the links at the same time as creating the links.

adeason commented on lib/AFS/CellCC/Dump.pm:
```diff
@@ -41,36 +41,51 @@ our @EXPORT_OK = qw(process_dumps);
 
+    for my $i (1 .. $#dump_fhs) {
+        if(-e $dump_fhs[$i]->filename) {
+            unlink($dump_fhs[$i]->filename)
```
just a note here: I was a little concerned that unlinking he files out from under File::Temp's files would do something weird (since we separately have a filehandle for the same file). It seems to be fine: File::Temp does have functionality to "safely delete" the underlying file (checking if it still points to the same file as we started with), but the default behavior doesn't really check for that.

I'm not complaining about it (I don't see any better way of doing this), but it's just worth being aware of that this is a little weird.

adeason commented on lib/AFS/CellCC/Dump.pm:
```diff
@@ -354,6 +374,79 @@ process_dumps($$$@) {
         # our caller.
         exit(1);
     }
+    $pm->wait_all_children();
```
this isn't quite what I meant when I said before we could do the _calc_incremental call, and then wait for the children to finish. By "waiting for them to finish" I meant, we should examine the state of the jobs to see if they are done running, and start running dumps when they're done. By waiting for the actual child procs to finish, we stall the whole dump-server until they're all done; so if something gets stuck for any reason (e.g. a `vos` command hanging), we don't start any new dumps. That's not good.

Instead what I suggest here is having the `_calc_incremental` calls be done in a new state, and then we can check to see if any such jobs are still running in that state for that volume.

adeason reviewed (changes requested):
As mentioned elsewhere, I had some issues with the approaches taken wrt waiting for _calc_incremental children to finish, and with some functions handling multiple jobs at once. The _calc_incremental stuff is mentioned inline.

The multi-jobs thing, though, I've gone back and forth on. The potential problems here are that if these jobs go into ERROR state, the checkserver will put them back into DUMP_START (or whatever) one at a time, so it's possible we'll retry only some of the jobs, instead of doing them all as one "batch". In addition, if the sync fails during the dumping stage, each job will generate an alert (even though we maybe only had a problem with one thing failing repeatedly). But these aren't so bad, no worse than annoyances; the sync will still work as they always did, and we're not worse off.

The potential alternative I was thinking of is to have just one job in e.g. DUMP_WORK, and have the others in some state like DUMP_WAIT. The DUMP_WORK job would create the dumpfile links for them, and transition the DUMP_WAIT jobs into DUMP_DONE when it transitions itself to DUMP_DONE. But that approach makes it more difficult to handle errors; if the _WORK job fails due to multiple errors, the _WAIT jobs won't go to ERROR state (unless we add logic to handle that). So that seems more error prone and your way is better; I just mention this to explain what I was whining about, and in case you had any thoughts on it.

For the comments on _calc_incremental scheduling, and the tempfile handling, I have an implementation I'd like you to look at, to see if you agree with the approach. I'll be pushing that into this pull request in a moment assuming I can figure it out and github lets me.

adeason commented:
> No functional difference should be incurred.

I know what you're trying to say, but when I see language like this, I read it to mean that what the code actually does haven't changed (like changing 'foo = foo+1' to 'foo++', or splitting code out into a separate function). That last commit does have differences, because it changes what's in log messages.

The commit is addressing errors in the earlier commits in this PR, though, so I don't think it should be a separate commit; just change the incorrect commits. I'll do that in a moment.

adeason commented:
For the PR history: Marcio and I discussed these changes earlier today, and he gave approval for the modified commits (except for the extra change he added, which I've incorporated). So I'm going ahead and merging this.
